src/domains/access.ml

```diff
@@ -421,35 +421,31 @@ end
 module LV = Printable.Prod (CilType.Varinfo) (O)
 module LVOpt = Printable.Option (LV) (struct let name = "NONE" end)
 
-
-let check_accs (prev_r,prev_lp,prev_w) (conf,w,loc,e,lp) =
-  match prev_r with
-  | None ->
-    let new_w  = prev_w || w in
-    let new_lp = LSSet.inter lp prev_lp in
-    let union_empty = LSSet.is_empty new_lp in
-    (* ignore(printf "intersection with %a = %a\n" LSSet.pretty lp LSSet.pretty new_lp); *)
-    let new_r  = if union_empty && new_w then Some conf else None in
-    (new_r, new_lp, new_w)
-  | _ -> (prev_r,prev_lp,prev_w)
+(* Check if two accesses race and if yes with which confidence *)
+let conflict2 (conf,mhp,w,loc,e,lp) (conf2,mhp2,w2,loc2,e2,lp2) =
+  if (not w) && (not w2) then
+    None (* two read/read accesses do not conflict *)
+  else if not (LSSet.is_empty @@ LSSet.inter lp lp2) then
+    None (* the labelled string set excludes that these conflict *)
+  else if not (MHP.may_happen_in_parallel mhp mhp2) then
+    None (* They may not be in parallel *)
+  else
+    Some (max conf conf2)
 
 let check_safe ls accs prev_safe =
   if ls = None then
     prev_safe
   else
-    let accs = AS.elements accs in (* hope that it is not nil *)
-    let lp_start = (fun (_,_,_,_,lp) -> lp) (List.hd accs) in
-    (* ignore(printf "starting with lockset %a\n" LSSet.pretty lp_start); *)
-    match List.fold_left check_accs (None, lp_start, false) accs, prev_safe with
-    | (None, _,_), _ ->
-      (* ignore(printf "this batch is safe\n"); *)
+    let accs = AS.elements accs in
+    let cart = List.cartesian_product accs accs in
+    let conf = List.filter_map (fun (x,y) -> if A.compare x y <= 0 then conflict2 x y else None) cart in
+    if conf = [] then
       prev_safe
-    | (Some n,_,_), Some m ->
-      (* ignore(printf "race with %d and %d \n" n m); *)
-      Some (max n m)
-    | (Some n,_,_), None ->
-      (* ignore(printf "race with %d\n" n); *)
-      Some n
+    else
+      let maxconf = List.max conf in
+      match prev_safe with
+      | None -> Some maxconf
+      | Some prev -> Some (max maxconf prev)
```

Comment:
Looks fine to me, although maybe @vesalvojdani should also have a look before we change this critical race detection code.